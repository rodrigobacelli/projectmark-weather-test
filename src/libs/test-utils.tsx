import * as React from 'react';
import { render, RenderOptions } from '@testing-library/react';

const Providers = ({ children }: { children: React.ReactNode }) => {
  return children;
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: Providers, ...options });

export * from '@testing-library/react';
export { customRender as render };
