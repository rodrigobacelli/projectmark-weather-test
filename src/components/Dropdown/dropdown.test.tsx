import { describe, it, expect, vi } from 'vitest';
import { render } from '../../tests/test-utils';
import { Dropdown } from './';

const mockedOptions = [
  {
    label: 'Option 1',
    value: 'option1',
  },
  {
    label: 'Option 2',
    value: 'option2',
  },
  {
    label: 'Option 3',
    value: 'option3',
  },
];

const mockedOptionsWithIcon = [
  ...mockedOptions,
  {
    label: 'Option 4',
    value: 'option4',
    icon: () => <div data-testid="option4-icon">Icon</div>,
  },
];

describe('components/Dropdown', () => {
  it('should handle selecting an option', async () => {
    const onSelect = vi.fn();

    const result = render(
      <Dropdown
        id="dropdown"
        className="dropdown-className"
        options={mockedOptions}
        selectedOption="option1"
        onSelect={onSelect}
      />
    );

    await result.getByRole('combobox').click();
    await result.getByText('Option 3').click();

    expect(onSelect).toHaveBeenCalledTimes(1);
    expect(onSelect).toHaveBeenCalledWith('option3');
  });

  it('should render icons for options with icons', async () => {
    const result = render(
      <Dropdown
        id="dropdown"
        className="dropdown-className"
        options={mockedOptionsWithIcon}
        selectedOption="option4"
      />
    );

    expect(
      result.getByRole('combobox').querySelector('[data-testid="option4-icon"]')
    ).toBeInTheDocument();

    await result.getByRole('combobox').click();

    expect(
      result.getByRole('listbox').querySelector('[data-testid="option4-icon"]')
    ).toBeInTheDocument();
  });

  it('should accept classNames', () => {
    const result = render(
      <Dropdown
        id="dropdown"
        className="dropdown-className"
        options={mockedOptions}
        selectedOption="option1"
      />
    );

    expect(
      result.container.querySelector('.dropdown-className')
    ).toBeInTheDocument();
  });

  it('should render the Header Component', () => {
    const result = render(
      <Dropdown
        id="dropdown"
        className="dropdown-className"
        options={[
          ...mockedOptions,
          {
            value: 'option4',
            label: 'Option 4',
          },
        ]}
        selectedOption="option1"
      />
    );

    expect(result).toMatchSnapshot();
  });
});
