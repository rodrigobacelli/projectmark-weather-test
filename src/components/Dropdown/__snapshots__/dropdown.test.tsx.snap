// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/Dropdown > should render the Header Component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative card-className"
      >
        <div
          aria-controls="dropdown-listbox"
          aria-expanded="false"
          aria-haspopup="dropdown-listbox"
          class="hover:cursor-pointer rounded-full px-4 py-2 border border-gray-800 dark:border-gray-200 hover:bg-gray-800/20 hover:dark:bg-gray-200/20"
          id="dropdown"
          role="combobox"
          tabindex="0"
        >
          <p
            class="text-gray-900 dark:text-gray-100 flex gap-2 items-center text-sm"
          >
            <span
              class="hidden sm:inline-flex"
            >
              Option 1
            </span>
            <svg
              class="ml-2 w-3 h-3"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              />
            </svg>
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative card-className"
    >
      <div
        aria-controls="dropdown-listbox"
        aria-expanded="false"
        aria-haspopup="dropdown-listbox"
        class="hover:cursor-pointer rounded-full px-4 py-2 border border-gray-800 dark:border-gray-200 hover:bg-gray-800/20 hover:dark:bg-gray-200/20"
        id="dropdown"
        role="combobox"
        tabindex="0"
      >
        <p
          class="text-gray-900 dark:text-gray-100 flex gap-2 items-center text-sm"
        >
          <span
            class="hidden sm:inline-flex"
          >
            Option 1
          </span>
          <svg
            class="ml-2 w-3 h-3"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
            />
          </svg>
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
